vdirsyncer/sync.py

```diff
@@ -99,107 +102,328 @@ def to_ident_conflict(self, storage):
                              hrefs=[self.old_href, self.new_href])
 
 
-class _Status(object):
-    def __init__(self, ident_to_props):
-        self._ident_to_props = ident_to_props
-        self._new_ident_to_props = {}
+class _StatusBase(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def transaction(self):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def insert_ident_a(self, ident, props):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def insert_ident_b(self, ident, props):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def update_ident_a(self, ident, props):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def update_ident_b(self, ident, props):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def remove_ident(self, ident):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_a(self, ident):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_b(self, ident):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_new_a(self, ident):
+        raise NotImplementedError()
 
-        self._href_to_status_a = dict((meta['href'], (ident, meta))
-                                      for ident, (meta, _)
-                                      in self._ident_to_props.items())
+    @abc.abstractmethod
+    def get_new_b(self, ident):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def iter_old(self):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def iter_new(self):
+        raise NotImplementedError()
 
-        self._href_to_status_b = dict((meta['href'], (ident, meta))
-                                      for ident, (_, meta)
-                                      in self._ident_to_props.items())
+    @abc.abstractmethod
+    def get_by_href_a(self, href, default=(None, None)):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_by_href_b(self, href, default=(None, None)):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def rollback(self, ident):
+        raise NotImplementedError()
+
+
+class _SqliteStatus(_StatusBase):
+    SCHEMA_VERSION = 1
+
+    def __init__(self, path):
+        self._path = path
+        self._conn = sqlite3.connect(path)
+        self._conn.row_factory = sqlite3.Row
+        self._c = None
+
+        if self._is_latest_version():
+            return
+
+        self._conn.execute('DROP TABLE IF EXISTS status')
```

Comment:
Technically, if you sync directly before upgrading vdirsyncer, it should be fine that your status data gets rebuilt from scratch. It is very slow though because it will fetch all items from both sides, and store them all in RAM (same thing happens on first usage of vdirsyncer).

I didn't think about this at all though, so now I replaced all DROP TABLE statements with a placeholder comment.